Mqtt processing stalled fix

tabascoz opened the pull request:
According to https://github.com/johanmeijer/grott/discussions/613 , this is the proposed fix for MQTT and Influxdb issues.

johanmeijer commented:
@tabascoz , I think something is going wrong here. You try to commit the alpha branche to the master.  That is not your intention (I suppose).

tabascoz commented:
Yep you're right, I changed it... Sorry for that :)

tabascoz commented:
Johan,

I did to test the theory and it's working fine, but I wasn't sure if having another py to deal with MQTT is necessary, but was unsure of where to put it. Also as the intention is to open a permanent connection to MQTT server and reuse that, where would be the best place to put.

johanmeijer commented:
Why I create the thread connection was to support multiple inverters sending a lot of data in a more stable way and to make the api more responsive.

I noticed that the get is really error prune/slow but to put works fine most of the time. That might be a prioritizing thing in the datalogger that we can not solve (or we have to create our own datalogger software it is an Esp.....))

I am curious to your latest test result and let see what can change.

tabascoz commented:
So latest results I reverted both influxdb and mqtt changes alone and find out that MQTT is actually the one triggering the issue, so no reason to change code for influxdb. 

The slower gets I believe you're right, nothing to do about it. I also think the same issue happens also if you change/get values from GRW web interface - you 'll see you get some not-so-rare failures. So it might not be particular to Grott. 

Also the issue i described I think there's no relation with it,  what I do to deal with it is to rely on a bash script that resubmits the put/get requests until i get an OK value. This error handling could be inserted in Grott and would help the functionality to be more reliable.  (perhaps I'll try to help and get it as a  next challenge.) 

I changed the code and inserted the class in grottdata, removed grottmqtt.py and changed it to  create and destroy mqtt connection object  just like the push_single does, so no more sharing of MQTT though threads. 

I read in paho that the library does not handle it well also, so it was definitely not a great idea. I just put the code of this PR to run and will report back. 

Pls just see that i included some more connections close request in the code, unsure if they are meaningful or not. 

Cheers.

tabascoz commented:
> Why I create the thread connection was to support multiple inverters sending a lot of data in a more stable way and to make the api more responsive.

Sorry forget to comment on that. The difference of it in my setup was tremendous. I strongly believe it's the direction to go for grott v3.

tabascoz commented:
@johanmeijer rock solid for the last 48 hours!!! All good. 

So stable that i see a time clock drift of some seconds in one of the inverters. Without constant inverter registrations, there are no more frequent time command sent.In 48hs it has drifted 30 seconds from the others. Nothing to worry about but i think a great  side effect :) 

So the threads looks fine no memory issues, connections to MQTT are all ok but the funny and weirdest thing was that i got "0" errors being caught by MQTT object. 

So here it goes another theory: I looked at the old log (you could give a look at the one i sent you) and looks all the time the issue started to happen it was when "concurrent" data has been sent by two inverters together. 

I was looking to Paho-MQTT and i think what is happening is that, as the library wasn't being made to deal with threads it has a reason as it does not create locks. So I think when the concurrent requests reach different inverter threads, it screws Paho-mqtt object. 

If that is true then, the issue would only show-up in a setup with multiple inverters + frequent updates ( and some time drift LoL) 

I think the theory explains it all and fits perfectly. Then another approach to the code i suggested would be to implement a lock mechanism on mqtt call. 

Cheers,

tabascoz commented:
I know you don't frequently take code contributions, but let me know if should I move  the PR to ready or close it. Cheers

johanmeijer commented:
Hi Fernando

 I am a little busy so I have not had the time to look at it yet.

Give me some time. I try to look at it the coming days.

johanmeijer commented:
@tabascoz I can not merge this request while it is still a draft request. 
 
I will merge it (see it as Grott version 3.2) and then see if this works for me also. To be safe I will also create a 3.1 branche to fall back to if necessary.

tabascoz commented:
hi 

> @tabascoz I can not merge this request while it is still a draft request.
> 
> I will merge it (see it as Grott version 3.2) and then see if this works for me also. To be safe I will also create a 3.1 branche to fall back to if necessary.

OK thanks, i changed it to ready. Let me know how it goes, so far so good here

tabascoz commented:
I also updated the title as for now  there's nothing to do with my initial suspicion of influxdb.

tabascoz commented:
> Thanks. I will commit it soon. I am travelling now. Please have some patience.

Sure! safe travels!
